component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -72,185 +73,325 @@
-                validateCustomClaims(claimsSet.getCustomClaims());
+        return true;
     }
 
-    /**
-     * Get the OauthAPP
-     *
-     * @param jwtSubject
-     * @return
-     * @throws IdentityOAuth2Exception
-     */
-    private OAuthAppDO getOAuthAppDO(String jwtSubject) throws IdentityOAuth2Exception {
-        OAuthAppDO oAuthAppDO = null;
-        try {
-            oAuthAppDO = OAuth2Util.getAppInformationByClientId(jwtSubject);
-        } catch (InvalidOAuthClientException e) {
-            handleException("Error while retrieving OAuth application with provided JWT information with subject:" +
-                    jwtSubject, e);
+    private Boolean validateWithClientId(String jwtClaim, String consumerKey) throws OAuthClientAuthnException {
+
+        if (consumerKey == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Unable to find OAuth application for provided JWT claim: " + jwtClaim);
+            }
+            throw new OAuthClientAuthnException("The issuer or the subject of the assertion is invalid.",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
         }
-        return oAuthAppDO;
+        if (!jwtClaim.trim().equals(consumerKey)) {
+            String errorMessage = "Invalid field: " + jwtClaim + " is found in the JWT. It should be equal to the: " +
+                    consumerKey;
+            if (log.isDebugEnabled()) {
+                log.debug(errorMessage);
+            }
+            throw new OAuthClientAuthnException("Invalid field: " + jwtClaim + " is found in the JWT", OAuth2ErrorCodes.
+                    INVALID_REQUEST);
+        }
+        return true;
     }
 
-    /**
-     * Retrieve the certificate and validate signature
-     *
-     * @param signedJWT
-     * @param tenantDomain
-     * @param alias
-     * @return
-     * @throws IdentityOAuth2Exception
-     */
-    public boolean isValidSignature(SignedJWT signedJWT, String tenantDomain,
-                                    String alias) throws IdentityOAuth2Exception {
-        try {
-            X509Certificate cert = getCertificate(tenantDomain, alias);
-            return validateSignature(signedJWT, cert);
-        } catch (JOSEException e) {
-            return handleException("Error when verifying signature with error:" + e.getMessage(), e);
+    // "The Audience SHOULD be the URL of the Authorization Server's Token Endpoint."
+    private boolean validateAudience(String tokenEP, List<String> audience) throws OAuthClientAuthnException {
+
+        for (String aud : audience) {
+            if (StringUtils.equals(tokenEP, aud)) {
+                return true;
+            }
+        }
+        String errorMessage = "None of the audience values matched the tokenEndpoint Alias :" + tokenEP;
+        if (log.isDebugEnabled()) {
+            log.debug(errorMessage);
         }
+        throw new OAuthClientAuthnException(errorMessage, OAuth2ErrorCodes.INVALID_REQUEST);
```

Comment:
Should we send the expected audience value to client. Isn't saying "audience doesn't match enough"